rainier-core/src/main/scala/com/stripe/rainier/core/Discrete.scala

```diff
@@ -121,32 +122,60 @@ final case class Poisson(lambda: Real) extends Discrete {
 
   val generator: Generator[Int] =
     Generator.require(Set(lambda)) { (r, n) =>
-      var lambdaLeft = n.toDouble(lambda)
+      val lambdaDouble = n.toDouble(lambda)
+      if (lambdaDouble < 30.0)
+        Poisson.small(lambdaDouble, r)
+      else
+        Poisson.large(lambdaDouble, r)
+    }
+
+  def logDensity(v: Real) =
+    lambda.log * v - lambda - Combinatorics.factorial(v)
+}
+
+object Poisson {
+  private def small(lambda: Double, r: RNG): Int = {
+    val l = math.exp(-lambda)
+    if (l >= 1.0) { 0 } else {
       var k = 0
       var p = 1.0
-
-      def update() = {
-        k = k + 1
+      while (p > l) {
+        k += 1
         p *= r.standardUniform
-        while (p < 1.0 && lambdaLeft > 0.0) {
-          if (lambdaLeft > Step) {
-            p *= eStep
-            lambdaLeft -= Step
-          } else {
-            p *= math.exp(lambdaLeft)
-            lambdaLeft = 0
-          }
-        }
       }
+      k - 1
+    }
+  }
 
-      update()
-      while (p > 1.0) update()
+  //from https://www.johndcook.com/blog/2010/06/14/generating-poisson-random-values/
+  private def large(lambda: Double, r: RNG): Int = {
+    val c = 0.767 - 3.36 / lambda
+    val beta = math.Pi / math.sqrt(3.0 * lambda)
+    val alpha = beta * lambda
+    val k = math.log(c) - lambda - math.log(beta)
 
-      k - 1
+    while (true) {
+      val u = r.standardUniform
+      val x = (alpha - math.log((1.0 - u) / u)) / beta
+      val n = math.floor(x + 0.5).toInt
+      if (n >= 0) {
+        val v = r.standardUniform
+        val y = alpha - beta * x
+        val lhs = y + math.log(v / math.pow(1.0 + math.exp(y), 2))
+        val rhs = k + n * math.log(lambda) - logFactorial(n)
+        if (lhs <= rhs)
+          return n
+      }
     }
 
-  def logDensity(v: Real) =
-    lambda.log * v - lambda - Combinatorics.factorial(v)
+    0
+  }
+
+  //from https://www.johndcook.com/blog/2010/08/16/how-to-compute-log-factorial/
+  private def logFactorial(n: Int): Double = {
```

Comment:
This form from the post has an error of `1/12(n+1)`, so at most ~0.04. Is that accurate enough for us? The post suggests using precomputed values for small numbers and then using this approximation (subtracting `1/(12 x)`) for large values.

Reply:
If we don't care about the extra accuracy, probably worth commenting about why

Reply:
Ok, I'll leave a comment. Empirically, there's a very small chance (around 0.1%) that `n` here will ever be < 10, even for a lambda=30 which is the lowest it can be, and that chance would go down for larger lambdas. So I think the bias here would be very slight.